tests/gnrc_tcp/tests/04-receive_data.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2018   Simon Brummer <simon.brummer@posteo.de>
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+import os
+import sys
+import threading
+
+from testrunner import run
+from shared_func import TcpServer, generate_port_number, get_host_tap_device, \
+                        get_host_ll_addr, get_riot_if_id, setup_internal_buffer, \
+                        read_data_from_internal_buffer, verify_pktbuf_empty
+
+
+def tcp_server(port, data):
+    with TcpServer(port) as tcp_srv:
+        tcp_srv.send(data)
+
+
+def testfunc(child):
+    port = generate_port_number()
+
+    # Try to receive 2000 bytes sent from the Host System.
+    data = '0123456789' * 200
+    data_len = len(data)
+
+    # Verify that RIOT Applications internal buffer can hold test data.
+    assert setup_internal_buffer(child) >= data_len
+
+    server_handle = threading.Thread(target=tcp_server, args=(port, data))
+    server_handle.start()
+
+    target_addr = get_host_ll_addr(get_host_tap_device()) + '%' + get_riot_if_id(child)
+
+    # Setup RIOT Node to connect to Hostsystems TCP Server
+    child.sendline('gnrc_tcp_tcb_init')
+    child.sendline('gnrc_tcp_open_active AF_INET6 ' + target_addr + " " + str(port) + ' 0')
+    child.expect_exact('gnrc_tcp_open_active: returns 0')
+
+    # Accept Data sent by the host system
+    child.sendline('gnrc_tcp_recv 0 ' + str(data_len))
+    child.expect_exact('gnrc_tcp_recv: received ' + str(data_len))
```

Comment:
This is not synchronized with `tcp_server` so you might run into a race condition were the data is not sent yet.

Reply:
(I get a lot of `gnrc_tcp_recv: returned -EAGAIN` here and only sometimes after that the expected message)

Reply:
Could be Nagles algorithm too.

Sending a packet from Linux does not necessarily mean that it is sent directly. Since gnrc_tcp_recv is called directly after sending it might be that the call tries to fetch data that was not sent before -> repeated -EAGAIN until the packet is actually sent. 

I'll see what I can find on the weekend.

Reply:
And not being synchronized is actually not a bad thing there. The entire point of TCP is to synchronize network transfer between systems.

Reply:
I have the same behavior (returns of -EAGAIN) while executing the test on a "nucleo-f401re".
I'll try to tweak the hosts socket options.

Reply:
Okay I "fixed" the issue. Actually the tests have shown perfectly valid behavior. By supplying no timeout gnrc_tcp_recv returns if there are was no data received.

Since Linux buffers TCP data to send internally instead of sending it right away, gnrc_tcp_recv  tries to fetch data that has not been sent before. I fixed the issue by supplying a timeout value forcing gnrc_tcp_recv to wait for incoming data.